src/librustc_target/abi/mod.rs

```diff
@@ -1081,4 +1119,57 @@ impl<'a, Ty> TyLayout<'a, Ty> {
             Abi::Aggregate { sized } => sized && self.size.bytes() == 0
         }
     }
+
+    /// Determines if zero-initializing this type might be okay.
+    /// This is conservative: in doubt, it will answer `true`.
+    pub fn might_permit_zero_init<C, E>(
+        &self,
+        cx: &C,
+    ) -> Result<bool, E>
+    where
+        Self: Copy,
+        Ty: TyLayoutMethods<'a, C>,
+        C: LayoutOf<Ty = Ty, TyLayout: MaybeResult<Self, Error = E>>
+    {
+        fn scalar_allows_zero(s: &Scalar) -> bool {
+            (*s.valid_range.start() <= 0) || // `&& *s.valid_range.end() >= 0` would be redundant
+            (*s.valid_range.start() > *s.valid_range.end()) // wrap-around allows 0
+        }
+
+        // Abi is the most informative here.
+        let res = match &self.abi {
+            Abi::Uninhabited => false, // definitely UB
+            Abi::Scalar(s) => scalar_allows_zero(s),
+            Abi::ScalarPair(s1, s2) =>
+                scalar_allows_zero(s1) && scalar_allows_zero(s2),
+            Abi::Vector { element: s, count } =>
+                *count == 0 || scalar_allows_zero(s),
+            Abi::Aggregate { .. } => {
+                // For aggregates, recurse.
+                let inner = match self.variants {
+                    Variants::Multiple { .. } => // FIXME: get variant with "0" discriminant.
+                        return Ok(true),
+                    Variants::Single { index } => self.for_variant(&cx, index),
+                };
+
+                match inner.fields {
+                    FieldPlacement::Union(..) => true, // An all-0 unit is fine.
+                    FieldPlacement::Array { .. } =>
+                        // FIXME: The widely use smallvec 0.6 creates uninit arrays
+                        // with any element type, so let us not (yet) complain about that.
+                        // count == 0 || inner.field(cx, 0).to_result()?.might_permit_zero_init(cx)?
+                        true,
+                    FieldPlacement::Arbitrary { ref offsets, .. } =>
+                        // Check that all fields accept zero-init.
+                        (0..offsets.len()).try_fold(true, |accu, idx|
```

Comment:
`try_fold` is one of those things that are mostly an optimization trick for internal iteration, this would probably be more readable as a plain loop.

Reply:
Hm, I actually like the functional immutable style here, but as you wish.

Reply:
Ideally this would be `.all` and then I would like that, but error-handling gets in the way...

Btw, I just remembered that `Err` is used with `try_fold` for early return, so your usage is unidiomatic, heh (here the return type would be `Result<(), Result<(), Error>>` with `Ok(())` meaning `true` and `Err(Ok(()))` meaning `false`).

Reply:
Early return is exactly what I want, so I do not see anything unidiomatic about it. As far as I know this is exactly what `try_fold` is for -- `all`/`any`-style iteration with support for error handling.

Reply:
You're using the "fold" part of `try_fold` for the `bool` result, but `all`/`any` *don't*, instead using the "early exit with error `()`" to indicate one of the values, and "normal successful exit with fold result `()`" to indicate the other, e.g.:
https://github.com/rust-lang/rust/blob/3f0e16473de5ec010f44290a8c3ea1d90e0ad7a2/src/libcore/iter/traits/iterator.rs#L1873-L1884

Where `LoopState` is a neat `Result`-like which works with `try_fold` because it implements `Try`:
https://github.com/rust-lang/rust/blob/3f0e16473de5ec010f44290a8c3ea1d90e0ad7a2/src/libcore/iter/mod.rs#L363-L384

Reply:
Well I am relying on short-circuiting, but yes iteration doesn't abort when it finds a `false`.

Still, I'd argue my use of `try_fold` is definitely idiomatic -- iteration with a fallible operation.